### PR TITLE
Implement .Net Standard 2.1 IAsyncEnumerable + add new build targets

### PR DESCRIPTION
**Pull request by Shane32**
This PR comprises the following changes:
* Changes existing IAsyncEnumerable definitions within .Net 4.5 or .Net 4.6 to match .Net Standard 2.1 signatures
* Imports MS backwards compatibility layer (bcl) for .Net Standard 2.0 projects
* Adds .Net Standard 2.1 and .Net Core 3.0 targets for linq2db project
* Adds .Net Core 3.0 target for test projects

Notes:
* A set of tests against SQL Server 2017 pass on my machine
* As of now, for .Net 4.5 / 4.6 projects, DisposeAsync() returns `Task` rather than `ValueTask`.
* The old `IAsyncEnumerable` and `IAsyncEnumerator` interfaces are still within the `LinqToDB.Async` namespace (for .Net 4.5 / 4.6 projects).  I cannot easily make them `internal` since some other public interfaces inherit from them.  I can move them to the `Compatibility` folder and move them into the `System.Collections.Generic` namespace, if you like.

Unless otherwise directed, I plan to add the following changes to this PR:
* Import `System.Threading.Tasks.Extensions` for .Net 4.5 / 4.6 projects, so that `ValueTask` can be utilized for `DisposeAsync` signatures, unifying the DisposeAsync signature across all targets
* Change `IAsyncDbConnection` and `IAsyncDbTransaction` to implement `IAsyncDisposable` (they have explicitly defined `Task DisposeAsync()` methods now)
* Add `ToAsyncEnumerable` method for .Net Standard 2.0+ projects to `LinqToDB.LinqExtensions` (map to `AsyncExtensions.AsAsyncEnumerable`)

Please review the code so far, and also let me know if you have any comments about the remaining proposed changes.  If you like, I can make the proposed additional changes in a separate PR.

### CODE REVIEW

File: Source/LinqToDB/DataProvider/SapHana/SapHanaProviderAdapter.cs [hunk 1]
```diff
@@ -15,9 +15,10 @@ public class SapHanaProviderAdapter : IDynamicProviderAdapter
 
 #if NET45 || NET46
 		public const string AssemblyName        = "Sap.Data.Hana.v4.5";
-#endif
-#if NETCOREAPP2_1
+#elif NETCOREAPP2_1
 		public const string AssemblyName        = "Sap.Data.Hana.Core.v2.1";
+#elif NETCOREAPP3_0
+		public const string AssemblyName        = "Sap.Data.Hana.Core.v3.0";
 #endif
```
Comment (Shane32) [inline Source/LinqToDB/DataProvider/SapHana/SapHanaProviderAdapter.cs:22]: Please review this change.

File: Source/LinqToDB/DataProvider/SapHana/SapHanaTools.cs [hunk 2]
```diff
@@ -109,11 +109,12 @@ static string DetectProviderName()
...
 			{
-#if !NETSTANDARD2_0
+#if !NETSTANDARD2_0 && !NETSTANDARD2_1
 				case SapHanaProviderAdapter.ClientNamespace:
 				case "Sap.Data.Hana.v4.5"                  :
 				case "Sap.Data.Hana.Core"                  :
 				case "Sap.Data.Hana.Core.v2.1"             :
+				case "Sap.Data.Hana.Core.v3.0"             :
 				case ProviderName.SapHanaNative            : return _hanaDataProvider.Value;
 #endif
```
Comment (Shane32) [inline Source/LinqToDB/DataProvider/SapHana/SapHanaTools.cs:119]: Please review this change

File: Source/LinqToDB/Linq/Expressions.cs [hunk 3]
```diff
@@ -527,7 +531,7 @@ public void SetInfo(MappingSchema mappingSchema)
 			{ M(() => "".Replace    (' ',' ') ), N(() => L<string?,char,char,string?>      ((string? obj,char   p0,char   p1)         => Sql.Replace  (obj, p0, p1))) },
 			{ M(() => "".Trim       ()        ), N(() => L<string?,string?>                ((string? obj)                             => Sql.Trim     (obj))) },
 
-#if NETCOREAPP2_1
+#if NETCOREAPP2_1 || NETCOREAPP3_0 //review
 			{ M(() => "".TrimEnd    ()        ), N(() => L<string,string?>                 ((string obj)                              =>     TrimRight(obj))) },
 			{ M(() => "".TrimStart  ()        ), N(() => L<string,string?>                 ((string obj)                              =>     TrimLeft (obj))) },
 #else
```
Comment (Shane32) [inline Source/LinqToDB/Linq/Expressions.cs:537]: Please review this change

File: Source/LinqToDB/Async/AsyncExtensions.cs [hunk 4]
```diff
@@ -61,9 +61,13 @@ public static IAsyncEnumerable<TSource> AsAsyncEnumerable<TSource>(this IQueryab
 			if (source == null) throw new ArgumentNullException(nameof(source));
 
 			var result = new List<T>();
-			using (var enumerator = source.GetEnumerator())
+#if NET45 || NET46
+			using (var enumerator = source.GetAsyncEnumerator(cancellationToken))
+#else
+			await using (var enumerator = source.GetAsyncEnumerator(cancellationToken))
```
Comment (MaceWindu) [inline Source/LinqToDB/Async/AsyncExtensions.cs:67]: need to attach `.ConfigureAwait(Common.Configuration.ContinueOnCapturedContext)` to all new `await` expressions

File: Source/LinqToDB/Async/AsyncExtensions.cs [hunk 5]
```diff
@@ -61,9 +61,13 @@ public static IAsyncEnumerable<TSource> AsAsyncEnumerable<TSource>(this IQueryab
 			if (source == null) throw new ArgumentNullException(nameof(source));
 
 			var result = new List<T>();
-			using (var enumerator = source.GetEnumerator())
+#if NET45 || NET46
+			using (var enumerator = source.GetAsyncEnumerator(cancellationToken))
```
Comment (MaceWindu) [inline Source/LinqToDB/Async/AsyncExtensions.cs:65]: I assume branching will be removed after IAsyncDisposable added to net45/net46 with ValueTask (or `await using` with work with Task-based version?)?

File: Source/LinqToDB/Async/AsyncExtensions.cs [hunk 6]
```diff
@@ -126,9 +130,13 @@ public static IAsyncEnumerable<TSource> AsAsyncEnumerable<TSource>(this IQueryab
 		{
 			if (source == null) throw new ArgumentNullException(nameof(source));
 
-			using (var enumerator = source.GetEnumerator())
+#if NET45 || NET46
+			using (var enumerator = source.GetAsyncEnumerator(cancellationToken))
+#else
+			await using (var enumerator = source.GetAsyncEnumerator(cancellationToken))
```
Comment (MaceWindu) [inline Source/LinqToDB/Async/AsyncExtensions.cs:136]: attach `.ConfigureAwait(Common.Configuration.ContinueOnCapturedContext)`

File: Source/LinqToDB/Async/AsyncExtensions.cs [hunk 7]
```diff
@@ -126,9 +130,13 @@ public static IAsyncEnumerable<TSource> AsAsyncEnumerable<TSource>(this IQueryab
 		{
 			if (source == null) throw new ArgumentNullException(nameof(source));
 
-			using (var enumerator = source.GetEnumerator())
+#if NET45 || NET46
+			using (var enumerator = source.GetAsyncEnumerator(cancellationToken))
```
Comment (MaceWindu) [inline Source/LinqToDB/Async/AsyncExtensions.cs:134]: `remove branching` candidate

File: Source/LinqToDB/Async/AsyncExtensions.cs [hunk 8]
```diff
@@ -146,9 +154,13 @@ public static async Task<TSource> FirstAsync<TSource>(this IAsyncEnumerable<TSou
 		{
 			if (source == null) throw new ArgumentNullException(nameof(source));
 
-			using (var enumerator = source.GetEnumerator())
+#if NET45 || NET46
+			using (var enumerator = source.GetAsyncEnumerator(token))
```
Comment (MaceWindu) [inline Source/LinqToDB/Async/AsyncExtensions.cs:158]: `remove branching` candidate

File: Source/LinqToDB/Async/AsyncExtensions.cs [hunk 9]
```diff
@@ -146,9 +154,13 @@ public static async Task<TSource> FirstAsync<TSource>(this IAsyncEnumerable<TSou
...
+#else
+			await using (var enumerator = source.GetAsyncEnumerator(token))
```
Comment (MaceWindu) [inline Source/LinqToDB/Async/AsyncExtensions.cs:160]: attach `.ConfigureAwait(Common.Configuration.ContinueOnCapturedContext)`

File: Source/LinqToDB/Async/AsyncExtensions.cs [hunk 10]
```diff
@@ -146,9 +154,13 @@ public static async Task<TSource> FirstAsync<TSource>(this IAsyncEnumerable<TSou
...
+#endif
 			{
-				if (await enumerator.MoveNext(token))
+				if (await enumerator.MoveNextAsync())
```
Comment (MaceWindu) [inline Source/LinqToDB/Async/AsyncExtensions.cs:163]: we had no `.ConfigureAwait(Common.Configuration.ContinueOnCapturedContext)` here before? Should be added

File: Source/LinqToDB/Async/IAsyncEnumerator.cs [hunk 11]
```diff
@@ -19,11 +19,15 @@ public interface IAsyncEnumerator<out T> : IDisposable
...
-		/// <param name="cancellationToken">Cancellation token that can be used to cancel the operation.</param>
 		/// <returns>
 		/// Task containing the result of the operation: true if the enumerator was successfully advanced
 		/// to the next element; false if the enumerator has passed the end of the sequence.
 		/// </returns>
-		Task<bool> MoveNext(CancellationToken cancellationToken);
+		Task<bool> MoveNextAsync();
+
+		/// <summary>
+		/// Disposes the object asynchonously
```
Comment (MaceWindu) [inline Source/LinqToDB/Async/IAsyncEnumerator.cs:29]: Xml-comments should end with dot

File: Source/LinqToDB/Async/IAsyncEnumerator.cs [hunk 12]
```diff
@@ -19,11 +19,15 @@ public interface IAsyncEnumerator<out T> : IDisposable
...
+		/// </summary>
+		Task DisposeAsync();
```
Comment (MaceWindu) [inline Source/LinqToDB/Async/IAsyncEnumerator.cs:31]: note: should be `ValueTask` after nuget added

File: Source/LinqToDB/Data/CommandInfo.cs [hunk 13]
```diff
@@ -675,27 +675,39 @@ class ReaderAsyncEnumerator<T> : IAsyncEnumerator<T>
...
 
-			public ReaderAsyncEnumerator(CommandInfo commandInfo, DbDataReader rd)
+			public ReaderAsyncEnumerator(CommandInfo commandInfo, DbDataReader rd, CancellationToken cancellationToken)
 			{
 				_commandInfo   = commandInfo;
 				_rd            = rd;
 				_additionalKey = commandInfo.GetCommandAdditionalKey(rd);
 				_objectReader  = GetObjectReader<T>(commandInfo.DataConnection, rd, commandInfo.DataConnection.Command.CommandText, _additionalKey);
 				_isFaulted     = false;
+				_cancellationToken = cancellationToken;
```
Comment (MaceWindu) [inline Source/LinqToDB/Data/CommandInfo.cs:687]: code alignment broken. Check https://marketplace.visualstudio.com/items?itemName=cpmcgrath.Codealignment  for formatting helper

File: Source/LinqToDB/Data/CommandInfo.cs [hunk 14]
```diff
@@ -675,27 +675,39 @@ class ReaderAsyncEnumerator<T> : IAsyncEnumerator<T>
...
 			}
 
 			public void Dispose()
 			{
 			}
 
+#if NET45 || NET46
+			public Task DisposeAsync() => TaskEx.CompletedTask;
```
Comment (MaceWindu) [inline Source/LinqToDB/Data/CommandInfo.cs:695]: `remove branching` candidate

File: Source/LinqToDB/Data/CommandInfo.cs [hunk 15]
```diff
@@ -675,27 +675,39 @@ class ReaderAsyncEnumerator<T> : IAsyncEnumerator<T>
...
+#else
+			public ValueTask DisposeAsync() => new ValueTask(Task.CompletedTask);
+#endif
+
 			public T Current { get; set; } = default!;
 
-			public async Task<bool> MoveNext(CancellationToken cancellationToken)
+#if NET45 || NET46
+			public async Task<bool> MoveNextAsync()
```
Comment (MaceWindu) [inline Source/LinqToDB/Data/CommandInfo.cs:703]: `remove branching` candidate

File: Source/LinqToDB/Data/DataConnection.QueryRunner.cs [hunk 16]
```diff
@@ -467,6 +467,20 @@ public void Dispose()
 					// methods for .net core 1x
 					DataReader.Dispose();
 				}
+
+#if NET45 || NET46
```
Comment (MaceWindu) [inline Source/LinqToDB/Data/DataConnection.QueryRunner.cs:471]: `remove branching` candidate

File: Build/linq2db.Source.props [hunk 17]
```diff
@@ -2,7 +2,7 @@
 	<Import Project="linq2db.Default.props" />
 
 	<PropertyGroup>
-		<TargetFrameworks>net45;net46;netstandard2.0;netcoreapp2.1</TargetFrameworks>
+		<TargetFrameworks>net45;net46;netstandard2.0;netcoreapp2.1;netstandard2.1;netcoreapp3.0</TargetFrameworks>
```
Comment (MaceWindu) [inline Build/linq2db.Source.props:5]: here and everywhere else: netcoreapp3.0/NETCOREAPP3_0 should be replaced with 3.1 as 3.0 is EOL already.

File: Source/LinqToDB/DataProvider/SapHana/SapHanaProviderAdapter.cs (hunk 1 above)
Comment (MaceWindu) [reply Source/LinqToDB/DataProvider/SapHana/SapHanaProviderAdapter.cs:22]: I don't think SAP HANA has this assembly, so it should continue using 2.1 (didn't checked their feb.2020 release, but doubting they added v3.x)

File: Source/LinqToDB/DataProvider/SapHana/SapHanaTools.cs (hunk 2 above)
Comment (MaceWindu) [reply Source/LinqToDB/DataProvider/SapHana/SapHanaTools.cs:119]: remove

File: Source/LinqToDB/DataProvider/SapHana/SapHanaProviderAdapter.cs [hunk 18]
```diff
@@ -15,9 +15,10 @@ public class SapHanaProviderAdapter : IDynamicProviderAdapter
 
 #if NET45 || NET46
 		public const string AssemblyName        = "Sap.Data.Hana.v4.5";
-#endif
-#if NETCOREAPP2_1
+#elif NETCOREAPP2_1
 		public const string AssemblyName        = "Sap.Data.Hana.Core.v2.1";
+#elif NETCOREAPP3_0
```
Comment (MaceWindu) [inline Source/LinqToDB/DataProvider/SapHana/SapHanaProviderAdapter.cs:20]: 3.0 -> 3.1

File: NuGet/linq2db.nuspec [hunk 19]
```diff
@@ -15,10 +15,19 @@
...
 			<group targetFramework=".NETCoreApp2.1">
 				<dependency id="System.ComponentModel.Annotations"    version="4.7.0" />
 				<dependency id="System.Data.DataSetExtensions"        version="4.5.0" />
+				<dependency id="Microsoft.Bcl.AsyncInterfaces"        version="1.1.1" />
+			</group>
+			<group targetFramework=".NETStandard2.1">
+				<dependency id="Microsoft.CSharp"                     version="4.7.0" />
+				<dependency id="System.ComponentModel.Annotations"    version="4.7.0" />
+			</group>
+			<group targetFramework=".NETCoreApp3.0">
```
Comment (MaceWindu) [inline NuGet/linq2db.nuspec:29]: 3.0 -> 3.1

File: Build/linq2db.Tests.props [hunk 20]
```diff
@@ -5,7 +5,7 @@
 	<PropertyGroup>
 	<!-- because VS doesn't support multi-targeting for tests (works for r# test runner): https://github.com/Microsoft/vstest/issues/298
 			 just put required test target framework to first position - VS always use first one only -->
-		<TargetFrameworks>net46;netcoreapp2.1</TargetFrameworks>
+		<TargetFrameworks>net46;netcoreapp2.1;netcoreapp3.0</TargetFrameworks>
```
Comment (MaceWindu) [inline Build/linq2db.Tests.props:8]: 3.0 -> 3.1

File: Build/linq2db.Tests.Providers.props [hunk 21]
```diff
@@ -70,7 +70,7 @@
 		</Content>
 	</ItemGroup>
 
-	<ItemGroup Condition=" '$(TargetFramework)' == 'netcoreapp2.1' ">
+	<ItemGroup Condition=" '$(TargetFramework)' == 'netcoreapp2.1' OR '$(TargetFramework)' == 'netcoreapp3.0' ">
```
Comment (MaceWindu) [inline Build/linq2db.Tests.Providers.props:73]: 3.0 -> 3.1

File: Source/LinqToDB/Linq/Expressions.cs [hunk 22]
```diff
@@ -271,6 +271,10 @@ public static void SetGenericInfoProvider(Type type)
 			var targetFramework = "netcoreapp2.1";
 #elif NETSTANDARD2_0
 			var targetFramework = "netstandard2.0";
+#elif NETCOREAPP3_0
+			var targetFramework = "netcoreapp3.0";
```
Comment (MaceWindu) [inline Source/LinqToDB/Linq/Expressions.cs:275]: 3.0 -> 3.1

File: Source/LinqToDB/Linq/Expressions.cs [hunk 23]
```diff
@@ -527,7 +531,7 @@ public void SetInfo(MappingSchema mappingSchema)
 			{ M(() => "".Replace    (' ',' ') ), N(() => L<string?,char,char,string?>      ((string? obj,char   p0,char   p1)         => Sql.Replace  (obj, p0, p1))) },
 			{ M(() => "".Trim       ()        ), N(() => L<string?,string?>                ((string? obj)                             => Sql.Trim     (obj))) },
 
-#if NETCOREAPP2_1
+#if NETCOREAPP2_1 || NETCOREAPP3_0 //review
```
Comment (MaceWindu) [inline Source/LinqToDB/Linq/Expressions.cs:534]: 3.0 -> 3.1

File: Source/LinqToDB/Linq/Expressions.cs (hunk 3 above)
Comment (MaceWindu) [reply Source/LinqToDB/Linq/Expressions.cs:537]: Should be correct. Those lines add support for new overloads, present only in netcore

File: Source/LinqToDB/Linq/IDataReaderAsync.cs [hunk 24]
```diff
@@ -6,6 +6,9 @@
 namespace LinqToDB.Linq
 {
 	public interface IDataReaderAsync : IDisposable
+#if !NET45 && !NET46
```
Comment (MaceWindu) [inline Source/LinqToDB/Linq/IDataReaderAsync.cs:9]: `remove ifded` candidate

File: Source/LinqToDB/Linq/QueryRunner.cs [hunk 25]
```diff
@@ -593,7 +593,12 @@ private static Type GetType<T>([DisallowNull] T obj, IDataContext db)
 		{
 			using (var runner = dataContext.GetQueryRunner(query, queryNumber, expression, ps, preambles))
 			{
-				using (var dr = await runner.ExecuteReaderAsync(cancellationToken).ConfigureAwait(Configuration.ContinueOnCapturedContext))
+				var dr = await runner.ExecuteReaderAsync(cancellationToken).ConfigureAwait(Configuration.ContinueOnCapturedContext);
+#if !NET45 && !NET46
+				await using (dr)
```
Comment (MaceWindu) [inline Source/LinqToDB/Linq/QueryRunner.cs:598]: attach `.ConfigureAwait(Common.Configuration.ContinueOnCapturedContext)`
also `remove branching` candidate

File: Source/LinqToDB/Linq/QueryRunner.cs [hunk 26]
```diff
@@ -648,30 +655,35 @@ class AsyncEnumeratorImpl<T> : IAsyncEnumerator<T>
 				_queryNumber = queryNumber;
 				_skipAction  = skipAction;
 				_takeAction  = takeAction;
+				_cancellationToken = cancellationToken;
```
Comment (MaceWindu) [inline Source/LinqToDB/Linq/QueryRunner.cs:658]: alignment

File: Source/LinqToDB/Linq/QueryRunner.cs [hunk 27]
```diff
@@ -648,30 +655,35 @@ class AsyncEnumeratorImpl<T> : IAsyncEnumerator<T>
...
 			}
 
 			public T Current { get; set; } = default!;
 
-			public async Task<bool> MoveNext(CancellationToken cancellationToken)
+#if NET45 || NET46
+			public async Task<bool> MoveNextAsync()
```
Comment (MaceWindu) [inline Source/LinqToDB/Linq/QueryRunner.cs:664]: `remove branching` candidate

File: Source/LinqToDB/Linq/QueryRunner.cs [hunk 28]
```diff
@@ -690,6 +702,12 @@ public void Dispose()
 
 				_queryRunner = null;
 			}
+
+#if NET45 || NET46
+			public Task DisposeAsync() => TaskEx.CompletedTask;
```
Comment (MaceWindu) [inline Source/LinqToDB/Linq/QueryRunner.cs:707]: `remove branching` candidate

File: Source/LinqToDB/Linq/QueryRunner.cs [hunk 29]
```diff
@@ -899,7 +917,12 @@ public IAsyncEnumerator<T> GetEnumerator()
 		{
 			using (var runner = dataContext.GetQueryRunner(query, 0, expression, ps, preambles))
 			{
-				using (var dr = await runner.ExecuteReaderAsync(cancellationToken).ConfigureAwait(Configuration.ContinueOnCapturedContext))
+				var dr = await runner.ExecuteReaderAsync(cancellationToken).ConfigureAwait(Configuration.ContinueOnCapturedContext);
+#if !NET45 && !NET46
+				await using (dr)
```
Comment (MaceWindu) [inline Source/LinqToDB/Linq/QueryRunner.cs:922]: attach `.ConfigureAwait(Common.Configuration.ContinueOnCapturedContext)`
also `remove branching` candidate

File: Source/LinqToDB/LinqToDB.csproj [hunk 30]
```diff
@@ -65,7 +65,7 @@
 		<Compile Include="Compatibility\System\Threading\AsyncLocal.cs" />
 	</ItemGroup>
 
-	<ItemGroup Condition=" '$(TargetFramework)' == 'netstandard2.0' OR '$(TargetFramework)' == 'netcoreapp2.1' ">
+	<ItemGroup Condition=" '$(TargetFramework)' == 'netstandard2.0' OR '$(TargetFramework)' == 'netcoreapp2.1' OR '$(TargetFramework)' == 'netstandard2.1' OR '$(TargetFramework)' == 'netcoreapp3.0' ">
```
Comment (MaceWindu) [inline Source/LinqToDB/LinqToDB.csproj:68]: 3.0 -> 3.1

File: Source/LinqToDB/LinqToDB.csproj [hunk 31]
```diff
@@ -75,16 +75,27 @@
...
+	</ItemGroup>
+
+	<ItemGroup Condition=" '$(TargetFramework)' == 'netstandard2.0' OR '$(TargetFramework)' == 'netstandard2.1' ">
 		<Compile Remove="DataProvider\SapHana\SapHanaDataProvider.cs" />
 		<Compile Remove="DataProvider\SapHana\SapHanaBulkCopy.cs" />
 
 		<PackageReference Include="Microsoft.CSharp" Version="4.7.0" />
 	</ItemGroup>
+	
+	<ItemGroup Condition=" '$(TargetFramework)' == 'netstandard2.1' OR '$(TargetFramework)' == 'netcoreapp3.0' OR '$(TargetFramework)' == 'netstandard2.0' OR '$(TargetFramework)' == 'netcoreapp2.1' ">
```
Comment (MaceWindu) [inline Source/LinqToDB/LinqToDB.csproj:94]: 3.0 -> 3.1

File: Source/LinqToDB/LinqToDB.csproj [hunk 32]
```diff
@@ -75,16 +75,27 @@
...
+		<Compile Remove="Async\IAsyncEnumerable.cs" />
+		<Compile Remove="Async\IAsyncEnumerator.cs" />
+	  <PackageReference Include="Microsoft.Bcl.AsyncInterfaces">
```
Comment (MaceWindu) [inline Source/LinqToDB/LinqToDB.csproj:97]: let's move version to attribute as we do for other imports

File: Tests/Base/TestBase.cs [hunk 33]
```diff
@@ -96,6 +96,8 @@ static TestBase()
 
 #if NETCOREAPP2_1
 			var configName = "CORE21";
+#elif NETCOREAPP3_0
+			var configName = "CORE30";
```
Comment (MaceWindu) [inline Tests/Base/TestBase.cs:100]: 3.0 -> 3.1

File: Tests/DataProviders.json [hunk 34]
```diff
@@ -48,6 +48,16 @@
 		]
 	},
 
+	"CORE30": {
```
Comment (MaceWindu) [inline Tests/DataProviders.json:51]: 3.0 -> 3.1

File: Tests/DataProviders.json [hunk 35]
```diff
@@ -60,6 +70,12 @@
 		"TraceLevel": "Info"
 	},
 
+	"CORE30.Azure": {
```
Comment (MaceWindu) [inline Tests/DataProviders.json:73]: 3.0 => 3.1

File: Tests/Linq/Tests.csproj [hunk 36]
```diff
@@ -15,7 +15,7 @@
 		<ProjectReference Include="..\FSharp\Tests.FSharp.fsproj" />
 	</ItemGroup>
 
-	<ItemGroup Condition=" '$(TargetFramework)' == 'netcoreapp2.1' ">
+	<ItemGroup Condition=" '$(TargetFramework)' == 'netcoreapp2.1' OR '$(TargetFramework)' == 'netcoreapp3.0' ">
```
Comment (MaceWindu) [inline Tests/Linq/Tests.csproj:18]: 3.0 => 3.1

File: Tests/Model/Tests.Model.csproj [hunk 37]
```diff
@@ -6,7 +6,7 @@
 		<AssemblyName>linq2db.Model</AssemblyName>
 	</PropertyGroup>
 
-	<ItemGroup Condition=" '$(TargetFramework)' == 'netcoreapp2.1' ">
+	<ItemGroup Condition=" '$(TargetFramework)' == 'netcoreapp2.1' OR '$(TargetFramework)' == 'netcoreapp3.0' ">
```
Comment (MaceWindu) [inline Tests/Model/Tests.Model.csproj:9]: 3.0 => 3.1

File: Tests/Tests.T4/Tests.T4.csproj [hunk 38]
```diff
@@ -134,7 +134,7 @@
 		</None>
 	</ItemGroup>
 
-	<ItemGroup Condition=" '$(TargetFramework)' == 'netcoreapp2.1' ">
+	<ItemGroup Condition=" '$(TargetFramework)' == 'netcoreapp2.1' OR '$(TargetFramework)' == 'netcoreapp3.0' ">
```
Comment (MaceWindu) [inline Tests/Tests.T4/Tests.T4.csproj:137]: 3.0 => 3.1

File: Tests/Tests.T4/Tests.T4.csproj [hunk 39]
```diff
@@ -304,7 +304,7 @@
 		<Reference Include="PresentationFramework" />
 		<Reference Include="WindowsBase" />
 	</ItemGroup>
-	<ItemGroup Condition=" '$(TargetFramework)' == 'netcoreapp2.1' ">
+	<ItemGroup Condition=" '$(TargetFramework)' == 'netcoreapp2.1' OR '$(TargetFramework)' == 'netcoreapp3.0' ">
```
Comment (MaceWindu) [inline Tests/Tests.T4/Tests.T4.csproj:307]: 3.0 => 3.1

**Comment by MaceWindu** [comment]
Please add new nugets to net45/net46 targets in separate commit after other changes, as we still need to decide if we want it

File: Source/LinqToDB/Async/AsyncExtensions.cs (hunk 5 above)
Comment (Shane32) [reply Source/LinqToDB/Async/AsyncExtensions.cs:65]: That's the idea.  I have not actually tried it yet.

File: Source/LinqToDB/Linq/QueryRunner.cs (hunk 25 above)
Comment (Shane32) [reply Source/LinqToDB/Linq/QueryRunner.cs]: Added ConfigureAwait.

File: Source/LinqToDB/Linq/QueryRunner.cs (hunk 29 above)
Comment (Shane32) [reply Source/LinqToDB/Linq/QueryRunner.cs]: Added ConfigureAwait.

**Comment by Shane32** [comment]
> As I mentioned before, I would prefer small PR without new targets for 3.0.0 (we still can do it), but let's try to finalize this one before release (we can always postpone it and release only breaking changes in 3.0.0).

Sorry; I must have misunderstood you.  I can certainly update the build configuration for the prior supported targets.  I've already updated all the `#if` conditions throughout for .Net Standard 2.1/.Net Core 3.1.  I can also make another branch and undo all those `#if` changes for 3.0.0 if you think that's wise.  Just let me know what you think is best at this point.

> > Add ToAsyncEnumerable method for .Net Standard 2.0+ projects to LinqToDB.LinqExtensions (map to AsyncExtensions.AsAsyncEnumerable)

> not sure what this means (why we need it)

The intent was to make the method a supported method, so consumers could asynchronously enumerate through a set of data, processing it as it becomes available from the data source.  Right now the method is in the LinqToDB.Async namespace, with a comment that says `This API supports the LinqToDB infrastructure and is not intended to be used directly from your code`.  I suggest moving or copying the method into the main namespace, probably inside the LinqExtensions.cs file, and making it a "supported" extension function.  What do you think?  This could be thought out after 3.0.0 is released, of course, as the function in question is already marked as an internal-use function.

> > The old IAsyncEnumerable and IAsyncEnumerator interfaces are still within the LinqToDB.Async namespace

> What stops us from just removing them and use proper interfaces?

The Microsoft.Bcl.AsyncInterfaces library is not available for .Net Framework 4.5 or .Net Framework 4.6 targets, so they cannot be removed.  Making them `internal` presents a complication, as `ILoadWithQueryable` and `IQueryProviderAsync` are public and rely on `IAsyncEnumerable`.

**Comment by Shane32** [comment]
I've implemented `IAsyncDisposable` for `IAsyncDbConnection` and related classes.  In the process, I realized that the .Net Standard 2.1 `DbConnection` class has asynchronous methods in it (e.g. `BeginTransactionAsync`), so the `AsyncDbConnection` class is probably not necessary anymore.  This seemed to be stretching past the scope of the original PR, so I did not make further changes besides the implementation of `IAsyncDisposable`.

See:
https://docs.microsoft.com/en-us/dotnet/api/system.data.common.dbconnection?view=netstandard-2.1
and
https://docs.microsoft.com/en-us/dotnet/api/system.data.common.dbtransaction?view=netstandard-2.1

Of note, the `IDbConnection` and `IDbTransaction` classes have not changed, and still do not have asynchronous members.

Here's a couple potential options to think about:
1. Make the `AsyncDbConnection` and `AsyncDbTransaction` classes inherit from the .Net Standard 2.1 counterparts.
2. Mark those classes and interfaces as deprecated for .Net Standard 2.1 projects, and convert any code that relies on those classes to use the .Net Standard 2.1 classes instead.
3. Eliminate the `AsyncDbConnection` and `AsyncDbTransaction` classes and interfaces entirely from the linq2db project for .Net Standard 2.1 targets.

**Comment by Shane32** [comment]
> Please add new nugets to net45/net46 targets in separate commit after other changes, as we still need to decide if we want it

I wrote the code but it doesn't run - lol - see #2304 for details.  All of the changes are in a single commit, as requested - see here:
* `Use ValueTask for .Net 4.5 and .Net 4.6 (broken)` https://github.com/linq2db/linq2db/pull/2304/commits/59cb4597e63c7f9d617456d40caf4949255757cb

File: Tests/Base/TestBase.cs [hunk 40]
```diff
@@ -96,6 +96,8 @@ static TestBase()
 
 #if NETCOREAPP2_1
 			var configName = "CORE21";
+#elif NETCOREAPP3_1
+			var configName = "CORE30";
```
Comment (MaceWindu) [inline Tests/Base/TestBase.cs:100]: CORE30 -> CORE31

**Comment by Shane32** [comment]
@MaceWindu I just noticed in another one of my projects that the `Microsoft.Bcl.AsyncInterfaces` nuget package is incompatible with the `System.Interactive.Async` nuget package prior to `4.0.0`.  This means anyone utilizing `System.Interactive.Async` 3.x _or any library dependent on this package_ (e.g. EF Core 2.x) would need to stick with linq2db 2.9.8, unless they are able to upgrade their packages to only utilize `System.Interactive.Async` 4.0.0 or newer.  I anticipate that you're fine with that, but if not, we could change the `#if` conditions to only support the new async interfaces for .Net Standard 2.1 projects.

https://github.com/dotnet/reactive

**Pull request merged**